File: src/parser_utils/package_exports.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Ast = Flow_ast

type t = { subpaths: Export_condition_map.t SMap.t }

let empty = { subpaths = SMap.empty }

let create ~subpaths = { subpaths }

let create_shorthand_index_exports ~export_path =
  let condition_map = Export_condition_map.create_from_shorthand ~path:export_path in
  let subpaths = SMap.empty |> SMap.add "." condition_map in
  create ~subpaths

let pattern_key_compare a b =
  let base_length_a =
    try String.index_from a 0 '*' with
    | Not_found -> -1
  in
  let base_length_b =
    try String.index_from b 0 '*' with
    | Not_found -> -1
  in
  if base_length_a > base_length_b then
    -1
  else if base_length_b > base_length_a then
    1
  else
    let length_a = String.length a in
    let length_b = String.length b in
    if length_a > length_b then
      -1
    else if length_b > length_a then
      1
    else
      0

let rec resolve_matched_wildcard_path subpath_map conditions match_key = function
  | expansion_key :: rest ->
    let pattern_base_index = String.index_from expansion_key 0 '*' in
    let pattern_base = String.sub expansion_key 0 pattern_base_index in
    if String.starts_with ~prefix:pattern_base match_key && match_key <> pattern_base then
      let pattern_trailer = Str.string_after expansion_key (pattern_base_index + 1) in
      if
        String.length pattern_trailer = 0
        || String.ends_with ~suffix:pattern_trailer match_key
           && String.length match_key >= String.length expansion_key
      then
        let target = SMap.find expansion_key subpath_map in
        let pattern_match =
          String.sub
            match_key
            (String.length pattern_base)
            (String.length match_key - String.length pattern_trailer - String.length pattern_base)
        in
        Export_condition_map.resolve_package_target target (Some pattern_match) conditions
      else
        resolve_matched_wildcard_path subpath_map conditions match_key rest
    else
      resolve_matched_wildcard_path subpath_map conditions match_key rest
  | [] -> None

let resolve_wildcard_package exports subpath conditions =
  let expansion_keys =
    SMap.bindings exports.subpaths
    |> List.map (fun (key, _) -> key)
    |> List.filter (fun key -> String.contains key '*')
    |> List.sort pattern_key_compare
  in
  resolve_matched_wildcard_path exports.subpaths conditions subpath expansion_keys

let resolve_package exports subpath conditions =
  match SMap.find_opt subpath exports.subpaths with
  | Some condition_map -> Export_condition_map.resolve_package_target condition_map None conditions
  | None -> resolve_wildcard_package exports subpath conditions

(** Given a list of JSON properties, loosely extract the properties and turn it into a
    [Expression.t SMap.t]. We aren't looking to validate the file, and don't currently
    care about any non-literal properties, so we skip over everything else. *)
let extract_property map property =
  let open Ast in
  let open Expression.Object in
  match property with
  | Property
      ( _,
        Property.Init
          {
            key = Property.StringLiteral (_, { StringLiteral.value = key; _ });
            value = (_, value);
            _;
          }
      ) ->
    SMap.add key value map
  | _ -> map

let collect_export_paths input_path value subpaths_map =
  match Export_condition_map.parse value with
  | Some condition_map -> SMap.add input_path condition_map subpaths_map
  | None -> subpaths_map

let parse_package_exports_obj properties =
  let prop_map = List.fold_left extract_property SMap.empty properties in
  let subpaths = SMap.fold collect_export_paths prop_map SMap.empty in
  create ~subpaths

let parse_conditional_exports_main_sugar obj =
  match Export_condition_map.parse obj with
  | Some condition_map ->
    let subpaths = SMap.empty |> SMap.add "." condition_map in
    Some (create ~subpaths)
  | None -> None

let is_conditional_exports_main_sugar properties =
  let is_path_prop = function
    | Ast.Expression.Object.Property
        ( _,
          Ast.Expression.Object.Property.Init
            {
              key =
                Ast.Expression.Object.Property.StringLiteral
                  (_, { Ast.StringLiteral.value = key; _ });
              _;
            }
        ) ->
      String.starts_with ~prefix:"." key
    | _ -> false
  in
  not (List.exists is_path_prop properties)

let parse expr =
  match expr with
  | Ast.Expression.StringLiteral { Ast.StringLiteral.value : string; _ } ->
    Some (create_shorthand_index_exports ~export_path:value)
  | Ast.Expression.Object obj ->
    let open Ast.Expression.Object in
    if is_conditional_exports_main_sugar obj.properties then
      parse_conditional_exports_main_sugar expr
    else
      Some (parse_package_exports_obj obj.properties)
  | _ -> None

File: rust_port/crates/flow_parser_utils/src/package_exports.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::HashMap;
use std::ops::Deref;

use dupe::Dupe;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_parser::ast;
use flow_parser::ast::expression::ExpressionInner;
use flow_parser::loc::Loc;

use crate::export_condition_map::ExportConditionMap;

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct PackageExports {
    subpaths: HashMap<FlowSmolStr, ExportConditionMap>,
}

impl PackageExports {
    pub fn empty() -> Self {
        PackageExports {
            subpaths: HashMap::new(),
        }
    }

    pub fn create(subpaths: HashMap<FlowSmolStr, ExportConditionMap>) -> Self {
        PackageExports { subpaths }
    }

    fn create_shorthand_index_exports(export_path: FlowSmolStr) -> Self {
        let condition_map = ExportConditionMap::create_from_shorthand(export_path);
        let mut subpaths = HashMap::new();
        subpaths.insert(FlowSmolStr::new("."), condition_map);
        PackageExports::create(subpaths)
    }

    fn resolve_wildcard_package(
        &self,
        subpath: &str,
        conditions: &[FlowSmolStr],
    ) -> Option<FlowSmolStr> {
        let mut expansion_keys: Vec<&FlowSmolStr> = self
            .subpaths
            .keys()
            .filter(|key| key.contains('*'))
            .collect();
        expansion_keys.sort_by(|a, b| pattern_key_compare(a, b));

        resolve_matched_wildcard_path(&self.subpaths, conditions, subpath, &expansion_keys)
    }

    pub fn resolve_package(
        &self,
        subpath: &str,
        conditions: &[FlowSmolStr],
    ) -> Option<FlowSmolStr> {
        match self.subpaths.get(subpath) {
            Some(condition_map) => condition_map.resolve_package_target(None, conditions),
            None => self.resolve_wildcard_package(subpath, conditions),
        }
    }

    pub fn parse(expr: &ast::expression::Expression<Loc, Loc>) -> Option<Self> {
        match expr.deref() {
            ExpressionInner::StringLiteral { inner, .. } => Some(
                PackageExports::create_shorthand_index_exports(inner.value.dupe()),
            ),
            ExpressionInner::Object { inner, .. } => {
                if is_conditional_exports_main_sugar(&inner.properties) {
                    parse_conditional_exports_main_sugar(expr)
                } else {
                    Some(parse_package_exports_obj(&inner.properties))
                }
            }
            _ => None,
        }
    }
}

fn pattern_key_compare(a: &str, b: &str) -> std::cmp::Ordering {
    let base_length_a = a.find('*').map_or(-1, |i| i as i32);
    let base_length_b = b.find('*').map_or(-1, |i| i as i32);

    if base_length_a > base_length_b {
        std::cmp::Ordering::Less
    } else if base_length_b > base_length_a {
        std::cmp::Ordering::Greater
    } else {
        let length_a = a.len();
        let length_b = b.len();
        if length_a > length_b {
            std::cmp::Ordering::Less
        } else if length_b > length_a {
            std::cmp::Ordering::Greater
        } else {
            std::cmp::Ordering::Equal
        }
    }
}

fn resolve_matched_wildcard_path(
    subpath_map: &HashMap<FlowSmolStr, ExportConditionMap>,
    conditions: &[FlowSmolStr],
    match_key: &str,
    expansion_keys: &[&FlowSmolStr],
) -> Option<FlowSmolStr> {
    for expansion_key in expansion_keys {
        if let Some(pattern_base_index) = expansion_key.find('*') {
            let pattern_base = &expansion_key[..pattern_base_index];

            if match_key.starts_with(pattern_base) && match_key != pattern_base {
                let pattern_trailer = &expansion_key[pattern_base_index + 1..];

                if pattern_trailer.is_empty()
                    || (match_key.ends_with(pattern_trailer)
                        && match_key.len() >= expansion_key.len())
                {
                    if let Some(target) = subpath_map.get(*expansion_key) {
                        let pattern_match =
                            &match_key[pattern_base.len()..match_key.len() - pattern_trailer.len()];

                        return target.resolve_package_target(Some(pattern_match), conditions);
                    }
                }
            }
        }
    }
    None
}

/// Given a list of JSON properties, loosely extract the properties and turn it into a
/// [Expression.t SMap.t]. We aren't looking to validate the file, and don't currently
/// care about any non-literal properties, so we skip over everything else.
fn extract_property(
    map: &mut HashMap<FlowSmolStr, ast::expression::Expression<Loc, Loc>>,
    property: &ast::expression::object::Property<Loc, Loc>,
) {
    if let ast::expression::object::Property::NormalProperty(
        ast::expression::object::NormalProperty::Init {
            key: ast::expression::object::Key::StringLiteral((_loc, lit)),
            value,
            ..
        },
    ) = property
    {
        map.insert(lit.value.dupe(), value.clone());
    }
}

fn collect_export_paths(
    input_path: FlowSmolStr,
    value: &ast::expression::Expression<Loc, Loc>,
    subpaths_map: &mut HashMap<FlowSmolStr, ExportConditionMap>,
) {
    if let Some(condition_map) = ExportConditionMap::parse(value) {
        subpaths_map.insert(input_path, condition_map);
    }
}

fn parse_package_exports_obj(
    properties: &[ast::expression::object::Property<Loc, Loc>],
) -> PackageExports {
    let mut prop_map = HashMap::new();
    for property in properties {
        extract_property(&mut prop_map, property);
    }

    let mut subpaths = HashMap::new();
    for (input_path, value) in prop_map {
        collect_export_paths(input_path, &value, &mut subpaths);
    }

    PackageExports::create(subpaths)
}

fn parse_conditional_exports_main_sugar(
    obj: &ast::expression::Expression<Loc, Loc>,
) -> Option<PackageExports> {
    match ExportConditionMap::parse(obj) {
        Some(condition_map) => {
            let mut subpaths = HashMap::new();
            subpaths.insert(FlowSmolStr::new("."), condition_map);
            Some(PackageExports::create(subpaths))
        }
        None => None,
    }
}

fn is_conditional_exports_main_sugar(
    properties: &[ast::expression::object::Property<Loc, Loc>],
) -> bool {
    fn is_path_prop(property: &ast::expression::object::Property<Loc, Loc>) -> bool {
        if let ast::expression::object::Property::NormalProperty(
            ast::expression::object::NormalProperty::Init {
                key: ast::expression::object::Key::StringLiteral((_loc, lit)),
                ..
            },
        ) = property
        {
            lit.value.starts_with('.')
        } else {
            false
        }
    }

    !properties.iter().any(is_path_prop)
}
